The `site-setup` onboarding flow uses the site id when the user is purchasing a domain

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

When the signup framework needs to operate on an existing site it uses the `siteSlug` query param. This is used by at least the `site-setup` and `site-launch` flows.

When we are about to direct the user towards the `site-setup` flow, and the user is purchasing a domain, this PR uses the site's ID rather than its slug. This is because the url for the `site-setup` flow is being generated at a time _before_ domain purchase has been completed, and we have no idea whether the domain purchase will be successful or not. If it's successful the site's primary url might change to something new. Using the site's ID works around this issue because it doesn't change.

This fixes an issue where, depending on the payment method, a newly purchased domain will have been set as the primary domain by the time we land in the `site-setup` flow, producing a WSOD.

p1632898918267200-slack-C029SB8JT8S

* `site-setup` flow accepts a `siteId` query param
* If the user has added a domain to their cart, then we use `siteId` in the signup destination instead of `siteSlug`. The reason I haven't always used `siteId` is because users are used to seeing the slug in the url, not random numbers.
* If `siteId` query param is present, but we don't have data for that store in redux, then we fetch it before displaying the signup UI. This mimics what we're already doing for the `siteSlug` query param. This could be useful if the local redux state is lost in between page transitions for some reason.
* Some actions (like switching flows) clear the current site slug out of the dependency store, so we do the same for the site id.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Make sure to test locally or on staging because `site-setup` is disabled in production. Or use the `signup/setup-site-after-checkout` feature flag.

Try various payment methods (e.g. CC, credits, paypal) and flows (e.g. `site-setup` and `site-launch`)

I've found that after completing `site-setup` you will land on `/home/{{ siteId }}` instead of the usual `/home/{{ siteSlug }}`. It still works fine though, and after navigating it switches to the site slug.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

p-jackson committed:
Use site id for site-setup flow when purchasing a domain

p-jackson committed:
Load site slug in dependencyStore when siteId query param is used

p-jackson committed:
Load siteId if it doesn't exist in redux store on page load

p-jackson committed:
Comment

arthur791004 commented on client/lib/signup/flow-controller.ts:
```diff
@@ -123,6 +124,22 @@ export default class SignupFlowController {
+		) {
+			const siteSlug = getSiteSlug(
+				this._reduxStore.getState(),
+				options.providedDependencies[ 'siteId' ]
```
If we already have `siteId`, can we fetch `siteSlug` by `siteId`? Or is there any problem? I feel like using `siteId` as `siteSlug` might be confused easily.

p-jackson replied:
Yes that's what the `getSiteSlug()` is doing here, we're fetching `siteSlug` by `siteId`. I might not quite understand what you mean though.

arthur791004 replied:
Oops...I saw `siteId` there and misunderstood this one 😅

arthur791004 commented on client/signup/controller.js:
```diff
@@ -331,24 +331,28 @@ export default {
 		} else {
-			// Fetch the site by siteSlug and then try to select again
-			dispatch( requestSite( siteSlug ) )
+			// Fetch the site by siteSlugOrId and then try to select again
```
`siteIdOrSlug` rather than ~siteSlugOrId~ 🤔

p-jackson replied:
Whoops! I changed by mind on the variable name half way through but forgot to update everywhere.
[Selectors like this one](https://github.com/Automattic/wp-calypso/blob/abdf35662927f9919535b675af4b2b2c7bcf5df4/client/state/sites/selectors/get-site-id.js#L18) call it `siteIdOrSlug` so I switched to using that order.

p-jackson committed:
Fix comment

arthur791004 reviewed (approved):
Test with credits and it looks good 👍

roo2 reviewed (changes requested):
It looks like when we're not purchasing a domain, and selecting a free plan, the page is throwing an error because it is not being passed `siteId` in this case (only `siteSlug`)

arthur791004 commented:
Yeah, you're right! I also encounter this problem!

p-jackson committed:
Make `siteId` an optional query param

`siteSlug` is still required, because we do need access to it. However
if `siteId` is provided then the correct `siteSlug` will be loaded into
the dependencies by the time the flow controller asserts for required
dependencies.

p-jackson commented:
@roo2 @arthur791004 I've been caught out by a Slack outage :( https://status.slack.com/ I can't properly see everything from the thread, could you paste anything that's been said into here?

p-jackson committed:
setSelectedSiteForSignup waits till the site is loaded before calling next()

roo2 reviewed (approved):
OK, retested and LGTM. Nice job fixing the issue with site data being cleared by the calypso initial state rehydration clearing dev chaos monkey thing.
